### ISSUE TITLE
Code Cleanup / Refactor

### ISSUE THREAD

**Issue by jekkos** [opensourcepos/opensourcepos#2339]
Just logging this issue as we should not forget to cleanup code further as discussed after merging.
- [ ] For the tax views, empty arrays are used instead of objects. More consistent would be to use empty objects as we do in other models and cast to an array
- [ ] `check4TaxCategoryDups`, validateTaxCategoryCharacters return statements can be a oneliner
- [ ] `requireTaxCategory` might not be needed, a required rule should be sufficient
- [ ] validators for tax rules should be added by using css class names instead of doing this in JS (which is shorter)
- [ ]  validator definitions itself should be extracted in a common view

**Comment by odiea** [comment]
@jekkos any idea when these updates will get to the demo server?

**Comment by jekkos** [comment]
Ok I changed the deploy script to now pull everytime from docker hub. It's on demo now.

**Comment by objecttothis** [comment]
@jekkos, should we add in here the refactoring of query() use into Query Builder and potentially the Database structure changes to reduce size and increase performance?  It's not urgent for 3.3.0, so maybe they go in a section of 4.0.0 or 3.4.0?

**Comment by daN4cat** [comment]
@SteveIreland due to the fact that above issues are related to Tax, is that something you addressed?

I wonder if this is really 3.3.0 or should be 3.4.0.

**Comment by daN4cat** [comment]
Moved to 3.4.0.

**Comment by daN4cat** [comment]
@jekkos how much of above checklist is still valid?

**Comment by jekkos** [comment]
As far as i know this is still outstanding technical debt.

**Comment by jekkos** [comment]
I was investigating some issue today, namely when changing the ospso_sales_items_taxes tax amount, it does not have any effect on the invoice.

It seems that this data is all saved in ospos_sales_tax table now. So I wonder, is the ospos_sales_item_taxes still used? @SteveIreland in this case you might have the best knowledge here. I think I remember we migrated away from this table? It does seem to be filled with correct values so I think at least that code could be cleaned up. (point to add here).

**Comment by SteveIreland** [comment]
@jekkos The sales_item_taxes table is still used for reporting.  It's intended to provide more accurate sales tax reporting.  When it was introduced it solved a calculation problem or two.  At the time I didn't think anyone would appreciate me touching the sales_tax table.

Regarding  "For the tax views, empty arrays are used instead of objects. More consistent would be to use empty objects as we do in other models and cast to an array." I'd love to address that issue if I understood what it was.  Can you provide an example of what is wrong and an example of how it should be?  I'm hoping you can point me to an example in your code that is "how it should be" and an equivalent segment in my code of how it should not be".    I spent a little bit of time on it and simply don't see the problems.  I know when I did the original development I was following examples from existing code thinking that would keep me out of trouble.  Apparently I picked the wrong examples.

**Comment by SteveIreland** [comment]
@jekkos You can continue to ignore me.  Once I started making changes I had my "oh duh" moment.

**Comment by jekkos** [comment]
@SteveIreland Hi Steve I was on holidays and not near a screen or keyboard, it's been a while since I logged this issue. I think it was more about keeping the code consistent then anything else, don't really prefer one way over the other, just consistent pattern is good enough.

**Comment by objecttothis** [comment]
> @jekkos The sales_item_taxes table is still used for reporting. It's intended to provide more accurate sales tax reporting. When it was introduced it solved a calculation problem or two. At the time I didn't think anyone would appreciate me touching the sales_tax table.
> 
> Regarding "For the tax views, empty arrays are used instead of objects. More consistent would be to use empty objects as we do in other models and cast to an array." I'd love to address that issue if I understood what it was. Can you provide an example of what is wrong and an example of how it should be? I'm hoping you can point me to an example in your code that is "how it should be" and an equivalent segment in my code of how it should not be". I spent a little bit of time on it and simply don't see the problems. I know when I did the original development I was following examples from existing code thinking that would keep me out of trouble. Apparently I picked the wrong examples.

As I've been converting the code base to CI 4 I have noticed that in some places we are using objects and in others we are using arrays. I agree that it would provide benefit to convert all of them to be the same. I don't have a preference as to which.  However, I do think this is better to do after the CI4 conversion is completed rather than have to do it twice.

**Comment by objecttothis** [comment]
> @objecttothis The last CI4 commit was December 10. Do you have a big one coming, or is the current state of the master branch ready to start testing? Sorry if it's a stupid question - I'm still catching up.

Not a stupid question at all.  PR #3592 has all the details of what still needs to be done.  I was on vacation for much of December and this last week.  I'm planning on getting back into the code tomorrow since my Wednesdays are reserved for another project.  I'm not sure about the state of the master because I've been focused on the ci4-upgrade branch. besides the checklist of things in that PR, right now I'm just trying to launch the webapp and fix what isn't working.  There are some issues with routes that are causing problems.  I could use some help really, but everyone is volunteer so it's based on people's willingness and availability.  Once we get to a point where the app loads and builds I think @odiea expressed a willingness to do regression testing.

If people start committing a bunch to the master, it's going to make a nightmare for merging that into the ci4-upgrade branch, so if people are looking for things to do, then I'd say helping finish off ci4-upgrade is my vote.

**Comment by odiea** [comment]
@objecttothis Yes I have been waiting to build this beast but have not been able to do so. Will wait for further instructions.

**Comment by objecttothis** [comment]
> @objecttothis Yes I have been waiting to build this beast but have not been able to do so. Will wait for further instructions.

I think getting weblate in shape and fixing the route problems are likely the only two things that are in the way at this point. I could use some help with both.  It's possible that the latest CI4 will fix some of the problems I am facing with routes, so tomorrow I will be working on incrementally upgrading from 4.1.3 to 4.3.1.

**Comment by SteveIreland** [comment]
After you can commit it, I will be participating in resolving any issues that are within my skill level.

For my customizations, I'm working on a branch off of the 3.3.8 release which is only in my fork- so I'm totally out of your way.
